metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/execution/WorkflowExecutor.java

```diff
@@ -170,36 +200,95 @@ private Date runInqueueOrRunningStateWorkflowExecution() {
     } else {
       finishDate = null;
     }
-    return finishDate;
+    return new ImmutablePair<>(finishDate, didPluginRun);
+  }
+
+  private int getFirstPluginPositionToStart() {
+    int firstPluginPositionToStart = 0;
+    for (int i = 0; i < workflowExecution.getMetisPlugins().size(); i++) {
+      AbstractMetisPlugin metisPlugin = workflowExecution.getMetisPlugins().get(i);
+      if (metisPlugin.getPluginStatus() == PluginStatus.INQUEUE
+          || metisPlugin.getPluginStatus() == PluginStatus.RUNNING
+          || metisPlugin.getPluginStatus() == PluginStatus.CLEANING
+          || metisPlugin.getPluginStatus() == PluginStatus.PENDING) {
+        firstPluginPositionToStart = i;
+        break;
+      }
+    }
+    return firstPluginPositionToStart;
+  }
+
+  /**
+   * Tries to acquire a semaphore permission corresponding to the provided plugin's type.
+   * <ol>
+   *   <li>If semaphore permission granted then there is space for that plugin and the plugin
+   *   starts</li>
+   *   <li>If semaphore permission NOT granted then the plugin din not run and a false flag is
+   *   send back as a return result</li>
+   * </ol>
+   *
+   * @param i the index of the plugin in the list of plugins inside the workflow execution
+   * @param plugin the provided plugin to be ran
+   * @return true if plugin ran, false if plugin did not run
+   * @throws InterruptedException if an interruption occurred
+   */
+  private boolean runMetisPluginWithSemaphoreAllocation(int i, AbstractMetisPlugin plugin)
+      throws InterruptedException {
+    // Sanity check
+    if (plugin == null) {
+      throw new IllegalStateException("Plugin cannot be null.");
+    }
+    // Check the plugin: it has to be executable
+    AbstractExecutablePlugin executablePlugin = expectExecutablePlugin(plugin);
+
+    final ExecutablePluginType executablePluginType = ExecutablePluginType
+        .getExecutablePluginFromPluginType(executablePlugin.getPluginType());
+    if (executablePluginType == null) {
+      throw new IllegalStateException("Plugin type cannot be null.");
+    }
+
+    //Try acquire semaphore and run plugin. Don't forget to release
+    boolean acquired = semaphoresPerPluginManager
+        .tryAcquireForExecutablePluginType(executablePluginType);
+    if (acquired) {
+      try {
+        //Update date for workflow execution so that it does not become stale
+        final Date updatedDate = new Date();
+        executablePlugin.setUpdatedDate(updatedDate);
+        workflowExecution.setUpdatedDate(updatedDate);
+        if (workflowExecution.getWorkflowStatus() != WorkflowStatus.RUNNING) {
+          workflowExecution.setStartedDate(updatedDate);
+          workflowExecution.setWorkflowStatus(WorkflowStatus.RUNNING);
+        }
+        workflowExecutionDao.updateMonitorInformation(workflowExecution);
```

Comment:
Are you sure you need to do this? I think this whole block won't be necessary (especially if the semaphore acquire returns immediately), and what's more, it interferes with some logic below (for instance line 302-304). 

If you really want to signal that the execution is not stale, perhaps just set the execution update time and leave it at that?

Reply:
The exact same thing is happening inside the claiming function.
But if the acquire timeout is removed then I guess this is not required here.